osquery/tables/cloud/ec2_instance_metadata.cpp

```diff
@@ -0,0 +1,283 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <string>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/network/protocol/http/client.hpp>
+#include <boost/noncopyable.hpp>
+#include <boost/property_tree/json_parser.hpp>
+
+#include <osquery/core.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+namespace pt = boost::property_tree;
+namespace http = boost::network::http;
+
+namespace osquery {
+namespace tables {
+
+/**
+ * @brief Super class for EC2 metadata accessors
+ */
+class Ec2MetaData : public boost::noncopyable {
+ protected:
+  /**
+   * @brief Meta data URL suffix
+   */
+  const std::string url_suffix_;
+
+  /**
+   * @brief Get metadata using HTTP.
+   *
+   * @return HTTP body.
+   */
+  std::string doGet() const;
+
+ public:
+  Ec2MetaData(const std::string urlSuffix) : url_suffix_(urlSuffix) {}
+
+  virtual ~Ec2MetaData() {}
+
+  /**
+   * @brief Extract relevant data from return API call, pure virtual
+   *
+   * @param http_body content of the http response body
+   * @param r The row to which the value need to be added
+   */
+  virtual void extractResult(const std::string& http_body, Row& r) const = 0;
+
+  /**
+   * @brief HTTP get and extract data
+   *
+   * @param r The row to which the value need to be added
+   */
+  void get(Row& r) const {
+    const std::string http_body = doGet();
+    extractResult(http_body, r);
+  }
+};
+
+/**
+ * @brief handle all data not requiring parsing
+ */
+class SimpleEc2MetaData : public Ec2MetaData {
+ protected:
+  /**
+   * @brief SQL type for the value
+   */
+  const ColumnType sql_type_;
+
+  /**
+   * @brief SQL column name
+   */
+  const std::string column_name_;
+
+ public:
+  SimpleEc2MetaData(const ColumnType sqlType,
+                    const std::string columnName,
+                    const std::string urlSuffix)
+      : Ec2MetaData(urlSuffix), sql_type_(sqlType), column_name_(columnName) {}
+
+  virtual void extractResult(const std::string& http_body,
+                             Row& r) const override;
+};
+
+/**
+ * @brief Handle IAM parsing
+ */
+class JSONEc2MetaData : public Ec2MetaData {
+ protected:
+  /**
+   * @brief SQL column names
+   */
+  const std::vector<std::string> column_names_;
+
+  /**
+   * @brief JSON key names
+   */
+  const std::vector<std::string> key_names_;
+
+ public:
+  JSONEc2MetaData(const std::vector<std::string> columnNames,
```

Comment:
For these parameters, is the intent to copy then move them into place? You may need to call `std::move` for the local ctor calls.